src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,261 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+            if (inputType.IsSZArray())
+            {
+                elementType = ((ArrayTypeSymbol)inputType).ElementType;
+            }
+            else if (TryFindIndexIndexerPatternForListPattern(node, inputType, out info, diagnostics))
+            {
+                elementType = info.Symbol.GetTypeOrReturnType().Type;
+            }
+            else
+            {
+                // PROTOTYPE(list-patterns) Cover any other type that we intend to cover, multi-dimensional arrays?
+                if (!hasErrors && !inputType.IsErrorType())
+                    Error(diagnostics, ErrorCode.ERR_UnsupportedTypeForListPattern, node, inputType);
+                hasErrors = true;
+                elementType = CreateErrorType();
+            }
+
+            ImmutableArray<BoundPattern> subpatterns = BindListPatternSubpatterns(node, inputType, elementType, permitDesignations, ref hasErrors, out bool sawSlice, diagnostics);
+            return new BoundListPatternClause(node, elementType, subpatterns, sawSlice, info, hasErrors);
+        }
+
+        private bool TryFindIndexIndexerPatternForListPattern(
+            SyntaxNode syntax, TypeSymbol receiverType,
+            [NotNullWhen(true)] out IndexerArgumentInfo? info,
+            BindingDiagnosticBag diagnostics)
+        {
+            using var _ = LookupResult.GetInstance(out var lookupResult);
+            return PerformPatternIndexerLookup(syntax, receiverType, lookupResult, out info, Compilation.GetWellKnownType(WellKnownType.System_Index), diagnostics) ||
+                   PerformPatternIndexerLookup(syntax, receiverType, lookupResult, out info, Compilation.GetSpecialType(SpecialType.System_Int32), diagnostics);
+        }
+
+        private bool TryFindRangeIndexerPatternForListPattern(
+            SyntaxNode syntax, TypeSymbol receiverType,
+            [NotNullWhen(true)] out IndexerArgumentInfo? info,
+            BindingDiagnosticBag diagnostics)
+        {
+            var useSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+            using var _ = LookupResult.GetInstance(out var lookupResult);
+            if (PerformPatternIndexerLookup(syntax, receiverType, lookupResult, out info, Compilation.GetWellKnownType(WellKnownType.System_Range), diagnostics))
+                return true;
+            if (!TryFindImplicitRangeIndexerPattern(syntax, receiverOpt: null, receiverType, out MethodSymbol? sliceMethod, BindingDiagnosticBag.Discarded, lookupResult, ref useSiteInfo))
+                return false;
+            info = new IndexerArgumentInfo(sliceMethod);
+            return true;
+        }
+
+        private bool PerformPatternIndexerLookup(
+            SyntaxNode syntax, TypeSymbol receiverType, LookupResult lookupResult,
+            [NotNullWhen(true)] out IndexerArgumentInfo? info, TypeSymbol? indexerArgumentType, BindingDiagnosticBag diagnostics)
+        {
+            info = null;
+
+            if (indexerArgumentType is null)
```

Comment:
Is it possible to get here with `null` `indexerArgumentType`?
Maybe this check should be for `indexerArgumentType` being an error type instead.

Reply:
Should not be possible to get null here. I just duplicated logic from elsewhere. Maybe an assert should be enough.
I think we should ensure there *is* a getter to begin with, otherwise we should have bailed earlier.

Reply:
Sorry I though you were commenting on `GetOwnOrInheritedGetMethod`

GetWellKnownType returns null in case it's missing, not an error type.

Reply:
I added an assertion here and none of the tests reach this branch. I think `Compilation.GetWellKnownType` never returns null (it returns error type symbols).

Reply:
You're right, I saw `result = null` and got fooled. it's later turned to an error type.